### ISSUE TITLE
MacOS Zsh Completion is unreasonably slow

### ISSUE THREAD

**Issue by UebelAndre** [bazelbuild/bazel#13578]
> ATTENTION! Please read and follow:
> - if this is a _question_ about how to build / test / query / deploy using Bazel, or a _discussion starter_, send it to bazel-discuss@googlegroups.com
> - if this is a _bug_ or _feature request_, fill the form below as best as you can.

### Description of the problem / feature request:

Related to https://github.com/bazelbuild/bazelisk/issues/29

I've installed bazelisk on a MacOS machine and am looking to use autocompletion in zsh and am experiencing painfully long delays between completions, up to 5 seconds for the initial tab and nearly 2 seconds for subsequent. It should be near instant.

### Feature requests: what underlying problem are you trying to solve with this feature?

> Replace this line with your answer.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Get a MacOS machine, brew install bazel/bazelisk and follow [completion guide](https://docs.bazel.build/versions/main/completion.html#zsh).

### What operating system are you running Bazel on?

MacOS 11.4 (20F71)

### What's the output of `bazel info release`?

release 4.1.0

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

> Replace this line with your answer.

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

> Replace this line with your answer.

###  Have you found anything relevant by searching the web?

> Replace these lines with your answer.
>
> Places to look:
> - StackOverflow: http://stackoverflow.com/questions/tagged/bazel
> - GitHub issues: https://github.com/bazelbuild/bazel/issues
> - email threads on https://groups.google.com/forum/#!forum/bazel-discuss

### Any other information, logs, or outputs that you want to share?

> Replace these lines with your answer.
>
> If the files are large, upload as attachment or provide link.

**Comment by loeffel-io** [comment]
+1

**Comment by sluongng** [comment]
I had sometimes to set this up today and I think majority of the slowness is due to the tab completion running a `bazel query` command underneath. For workspaces that have yet to perform a build, this would require user to fetch all the needed external dependencies and thus, making it slow.

One way to make it faster would be to narrow down the query scope using an environment variable config, OR to pre-generate the query result somewhere.

**Comment by meisterT** [comment]
I don't think we can do anything about the slowness, if we are invoking bazel query
